### PR TITLE
Add first pass at Docker support

### PR DESCRIPTION
**Pull request by rmccue**
See #193.

Not yet working. :( The problem is that everything generates upstart scripts, but phusion/baseimage uses runit. Apparently, upstart can't run inside Docker, so this is painful.

I suspect this can be resolved by upgrading to a newer base version of Ubuntu (16.04, e.g.), as it would use systemd instead of Upstart. That's a bigger topic, however.

### CODE REVIEW

**Comment by joehoyle** [comment]
Are you thinking we'll continue using Vagrant? Not sure if that's necessary, or if that makes an additional linux VM in VB?

**Comment by rmccue** [comment]
> Are you thinking we'll continue using Vagrant? Not sure if that's necessary, or if that makes an additional linux VM in VB?

Sticking with Vagrant provides a consistent CLI, and also ensures there's support for Docker even on systems without it with the VM. You can use the native Docker without the VM though if you have it installed (although that's not in this patch).

**Comment by joehoyle** [comment]
Ok, I would _really_ like to not install virtualbox / vmware if at all possible.

**Comment by rmccue** [comment]
See https://www.vagrantup.com/docs/docker/configuration.html#force_host_vm - Vagrant will use Docker natively without a VM if that's set to `false`

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# FROM ubuntu-upstart:precise
+# http://phusion.github.io/baseimage-docker/
+FROM phusion/baseimage:0.9.9
+
+# Install pre-pre-requisites.
+RUN apt-get update
```
Comment (BronsonQuick) [inline Dockerfile]: 🔢 Can we chain commands instead? e.g. `RUN apt-get update && apt-get install -q -y wget #ca-certificates`

Seeing I've been working with Docker on a client project I've been "levelling up" on Docker training today and the consensus I've seen is that Docker folks tend to prefer chaining as it leaves less overhead.

Comment (rmccue) [inline Dockerfile]: This isn't suitable for merge in any case, but best practice is to merge commands where it makes sense; it doesn't with update and install (IMO). Changing the command invalidates the intermediate container, but adding a new package shouldn't invalidate the `update`, usually.

Comment (BronsonQuick) [inline Dockerfile]: Yeah I know it's a super old PR and we'll have to update it. I came back and edited my comment from 🔢 as it only seems to make sense to chain `RUN` commands based on what Docker does behind the scenes.

I'm still doing a Docker course as we speak at the moment so I'm getting my head around all the different pieces as I'm keen to have a decent opinion about how Docker could work with Chassis and all our extensions. Long weekends FTW! I get to learn things!

Comment (rmccue) [inline Dockerfile]: Unfortunately, it can't: systemd can't run inside a container properly.

Comment (BronsonQuick) [inline Dockerfile]: Gotcha. When I learnt about Docker needing things to not be daemonized I figured that was probably a show stopper 😞

Comment (BronsonQuick) [inline Dockerfile]: For peeps playing along at home I'm incorrect with that comment! @rmccue mentioned this in our Slack.

>Docker can handle daemonisation fine
It's just systemd that doesn't like not having access to the root filesystem

**Comment by ntwb** [comment]
@rmccue Have you looked at https://github.com/ibuildthecloud/systemd-docker ?

**Comment by cstrouse** [comment]
Any thoughts on adding LXD support? That's probably a better fit for this job than Docker.

**Comment by BronsonQuick** [comment]
@cstrouse Oohh. I hadn't even heard of LXD before now. Looks like an interesting idea. I'll try and have a good read and experiment with it later this week and see if it's an option. Thanks for mentioning it!

**Comment by BronsonQuick** [comment]
I think I'll close this one as I can't see us getting around to this.

**Pull request closed without merging**